PostHog/Replay/URLSessionExtension.swift

```diff
@@ -0,0 +1,195 @@
+        } catch {
+            captureData(request: request, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    func postHogUpload(for request: URLRequest, from bodyData: Data) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await upload(for: request, from: bodyData)
+            after = Date()
+            captureData(request: request, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: request, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    func postHogData(for request: URLRequest, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await self.data(for: request, delegate: delegate)
+            after = Date()
+            captureData(request: request, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: request, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    func postHogData(from url: URL, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await self.data(from: url, delegate: delegate)
+            after = Date()
+            captureData(request: nil, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: nil, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    func postHogUpload(for request: URLRequest, fromFile fileURL: URL, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+        let before = Date()
+        var after: Date?
+        do {
+            let (data, response) = try await upload(for: request, fromFile: fileURL, delegate: delegate)
+            after = Date()
+            captureData(request: request, response: response, before: before, after: after)
+            return (data, response)
+        } catch {
+            captureData(request: request, response: nil, before: before, after: after)
+            throw error
+        }
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    func postHogUpload(for request: URLRequest, from bodyData: Data, delegate: (any URLSessionTaskDelegate)? = nil) async throws -> (Data, URLResponse) {
+        let before = Date()
```

Comment:
i assume this is the best clock to use? JS has "performance" clock that is (more) guaranteed to monotonically increase etc

Reply:
yes and no, https://developer.apple.com/documentation/swift/continuousclock
only from iOS >= 16 and we support iOS >= 13

Reply:
Let me check if there's anything available in C

Reply:
used something from the kernel, hopefully it works on all platforms https://developer.apple.com/documentation/kernel/1462446-mach_absolute_time